open-apparel-registry/open-apparel-registry
# Update facility download

## Overview

Updates the facility download format to split facility list items to
separate rows, with each facility having one base row and 0 or more
facility list item rows. In cases where a claim exists, the base row
is listed as contributed by the claimant; in other cases, the facility
created_from contributor is cited.

The facility CSV/XLSX download has become increasingly more complex
over time. Previously we used the facility search results
directly to download the list, but the additional data required to
render the download began to involve a lot of overhead, and we switched
to only downloading the more detailed version of the list when the
download was requested. With the latest changes to the format, the
required data was expected to again increase, in addition to the
complexity of the formatting increasing.

To simplify the process and decrease the amount of data fetched, the
formatting of the data into a CSV/XLSX-friendly format has been moved
to the backend.

Connects #1752 

## Demo

### CSV and XLSX files
[embed.csv](https://github.com/open-apparel-registry/open-apparel-registry/files/9153861/embed.csv)
[new_download.xlsx](https://github.com/open-apparel-registry/open-apparel-registry/files/9153863/new_download.xlsx)
[new_download.csv](https://github.com/open-apparel-registry/open-apparel-registry/files/9153865/new_download.csv)

### Timing

I timed the logDownload function locally while downloading 1108 facilities. This is a very rough estimate of timing, because we have significantly more list items per facility on staging and production, but the results are encouraging.

Time to download 1108 facilities on this branch: 36664.72119140625 ms (36.66 seconds)
Time to download 1108 facilities on develop: 97642.63012695312 ms (97.64 seconds)

## Notes

The previous front-end implementation of this included some sections which were wrapped in flags (embedded map, extended fields, PPE, claims, and closure status). It doesn't seem likely that we will remove embedded map, extended fields, claims, or closure status in the future, so I didn't wrap these portions of the code in flags. PPE is in the process of being phased out, so I didn't include PPE columns for download at all. 

I have it set up to filter out list items that are inactive, private, or have a status other than AUTOMATIC, CONFIRMED, or MERGED. Should some of these items instead be included but have the contributor anonymized? Are there are instances where the contributor should be anonymized?

## Testing Instructions

* Have the following types of facilities in your local database, each submitted at least once by the same contributor:
     - A claimed facility, unapproved
     - A claimed facility, approved
     - A facility list item with extended fields
     - A facility list item with nonstandard/contributor fields
     - An inactive or private facility list item
* Download search results including the given facilities as a CSV
* Download the same as XLSX
* Open the embedded map for the contributor and download the same facilities
* Confirm that each download contains the expected data and formatting
* View the new tests and confirm that the expected results for each test match your understanding of the expected download formatting

## Checklist
